### PR TITLE
[Mm 44864] Too long time stamp announcement

### PR DESCRIPTION
**Pull request by AbhinavVihan**
#### Summary
Time stamp announcement shortened (when focused on post time)

#### Ticket Link
fixes https://mattermost.atlassian.net/browse/MM-44864

#### Related Pull Requests
None

#### Screenshots
BEFORE -:

https://user-images.githubusercontent.com/89139362/190311992-f4e3101f-54dc-4470-ba34-8ed3d4946b3e.mp4

AFTER -:

https://user-images.githubusercontent.com/89139362/190312031-bb253432-e45f-4d2e-b3cf-fb0cf39e74a0.mp4

#### Release Note
NA
```release-note
Time stamp announcement shortened (when focused on post time)

```

### CODE REVIEW

**Review by nevyangelova — CHANGES_REQUESTED** [review]
Thanks @AbhinavVihan can you please update the failing snapshots?

**Comment by AbhinavVihan** [comment]
Done

**Comment by AbhinavVihan** [comment]
Hi @nevyangelova, Can you plaease take a look now.

**Review by nevyangelova — APPROVED** [review]

**Review by nevyangelova — APPROVED** [review]

**Comment by AshishDhama** [comment]
@jgilliam17 Gentle reminder on this one.

**Comment by AshishDhama** [comment]
@BenCookie95 Take a look.

**Review by BenCookie95 — CHANGES_REQUESTED** [review]
The ticket's title is a little misleading, I think the issue is that the timestamp announcement doesn't follow the users time display settings?

The announcement should match what appears in the tooltip when you hover over a posts timestamp I think. @jgilliam17 can you confirm?

**Comment by jgilliam17** [comment]
@BenCookie95 yes, I agree - as written in the ticket expected section, the announcement should follow user set time format i.e. 12h or 24h clock, and read out what is displayed for post time stamp

**Comment by AbhinavVihan** [comment]
@BenCookie95, So then length of announced timestamp should remain the same? **Reader announces the date and time using 24 hour clock, including adding GMT information**
or it should be shortened and announced in ***format selected by the user in settings for clock display***

**Comment by BenCookie95** [comment]
> @BenCookie95, So then length of announced timestamp should remain the same? **Reader announces the date and time using 24 hour clock, including adding GMT information** or it should be shortened and announced in _**format selected by the user in settings for clock display**_

I think the length should remain the same but if the timestamp has a time in there, the time should follow the setting from their clock display settings

**Comment by calebroseland** [comment]
@AbhinavVihan (and others) I'm sorry that I let this review slip. Albeit much late, here is some background and a few thoughts.

Ideally, I think that the announcement should match the contents of the tooltip, as the spoken announcement is an alternative presentation of the tooltip, which is a more precise timestamp—`time` and `date`.

CleanShot 2022-10-12 at 09 01 24@2x

IMO, shorting the `aria-label` it to just `time` negates the purpose of the more precise timestamp. It feels odd just having the time announced twice.

When Timestamp was first written, the `formatLabel` method was a bit of a shortcut and not polished. Additionally, the `datetime` attribute on `<time>` gives the screenreader what it needs (machine-readable ISO-time) to read out according to the screenreader's time locale options. 

## Possible Solution

To get the announcement to say what is in the tooltip, I think we should drop the default `area-label` from Timestamp/SemanticTime (SemanticTime would still be semantic because of the `datetime` attribute.) With that `aria-label` gone, it would let the tooltip's `labelled-by` attribute to point to the tooltip when focusing on the permalink. For the screenshot above, it would then announce: "Link, 5:15PM. September 15 at 5:15PM" 

I would suggest:
- in `timestamp`:
```diff
--- a/components/timestamp/timestamp.tsx
+++ b/components/timestamp/timestamp.tsx
@@ -429,7 +429,7 @@ class Timestamp extends PureComponent<Props, State> {
             formatted = (
                 <SemanticTime
                     value={value}
-                    aria-label={label ?? Timestamp.formatLabel(value, timeZone)}
+                    aria-label={label}
                     className={className}
                 >
                     {formatted}
```
- in `semantic_time`:
```diff
--- a/components/timestamp/semantic_time.tsx
+++ b/components/timestamp/semantic_time.tsx
@@ -1,6 +1,7 @@
 // Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
 // See LICENSE.txt for license information.
 
+import {DateTime} from 'luxon';
 import React, {FC, ReactNode, memo, TimeHTMLAttributes} from 'react';
 
 export type Props = {
@@ -11,14 +12,12 @@ export type Props = {
 const SemanticTime: FC<Props> = ({
     value,
     children,
-    'aria-label': label = value.toLocaleString(),
     ...props
 }: Props) => {
     return (
         <time
             {...props}
-            aria-label={label}
-            dateTime={value.toISOString()}
+            dateTime={DateTime.fromJSDate(value).toLocal().toISO({includeOffset: false})}
         >
             {children}
         </time>
```

- in `post_time`:
```diff
--- a/components/post_view/post_time/post_time.tsx
+++ b/components/post_view/post_time/post_time.tsx
@@ -101,6 +101,7 @@ export default class PostTime extends React.PureComponent<Props> {
                         <Timestamp
                             value={eventTime}
                             ranges={POST_TOOLTIP_RANGES}
+                            useSemanticOutput={false}
                         />
                     </Tooltip>
                 }
```


Again, I'm sorry for the long delay. Happy to continue to discuss.

**Comment by AbhinavVihan** [comment]
@calebroseland, Thankyou for all the information, truly.

**Review by BenCookie95 — CHANGES_REQUESTED** [review]
Thanks for looking into this again @AbhinavVihan but still a few issues. The announcer is now using the aria-describedby attribute from the link but it's still only reading the time and not any date information. We need to add `aria-labelledby={eventTime.toString()}` to the Link component in the post_time file and that will fix it!

Thanks for the detailed response @calebroseland!

File: components/post_view/post_time/post_time.tsx [hunk 1]
```diff
@@ -67,6 +67,7 @@ export default class PostTime extends React.PureComponent<Props> {
                 value={eventTime}
                 className='post__time'
                 useDate={false}
+                useSemanticOutput={false}
```
[the remark this answers is not in the document]
Comment (calebroseland) [reply components/post_view/post_time/post_time.tsx:70]: Yes, this should go onto the timestamp in the tooltip, not the one here.

**Review by BenCookie95 — APPROVED** [review]

**Review by calebroseland — APPROVED** [review]

**Comment by AbhinavVihan** [comment]
@jgilliam17, rebased with master

**Comment by jgilliam17** [comment]
Thanks for running e2es @M-ZubairAhmed 
[E2E report](https://mattermost-cypress-report.s3.amazonaws.com/272167-3bd3d58-onprem-ent-webapp-pr-11139-mm-ee-test:3bd3d58/mochawesome.html) looks good, no PR related failures.

**Review by jgilliam17 — APPROVED** [review]

**Pull request merged**